File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/median_of_two_sorted_arrays/MedianofTwoSortedArrays3.java
package main.java.com.jiang.leetcode100.median_of_two_sorted_arrays;

import java.util.Scanner;

class Solution3 {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        int m = nums1.length;
        int n = nums2.length;
        if((m + n) % 2 == 1) {
            return getKthElement(nums1, nums2, (m + n) / 2 + 1);
        } else {
            return (getKthElement(nums1, nums2, (m + n) / 2) + 
                getKthElement(nums1, nums2, (m + n) / 2 + 1)) / 2.0;
        }

    }

    public int getKthElement(int[] nums1, int[] nums2, int k) {
        int m = nums1.length;
        int n = nums2.length;
        int index1 = 0, index2 = 0;
        while(true) {
            if(index1 == m) return nums2[index2 + k - 1];
            if(index2 == n) return nums1[index1 + k - 1];
            if(k == 1) return Math.min(nums1[index1], nums2[index2]);

            int mid = k / 2;
            int newindex1 = Math.min(index1 + mid, m) - 1;
            int newindex2 = Math.min(index2 + mid, n) - 1;
            if(nums1[newindex1] <= nums2[newindex2]) {
                k -= (newindex1 - index1) + 1;
                index1 = newindex1 + 1;
            } else {
                k -= (newindex2 - index2) + 1;
                index2 = newindex2 + 1;
            }
        }
    }
}

public class MedianofTwoSortedArrays3 {
    public static void main(String[] args) {

        Scanner scan = new Scanner(System.in);
        String str1 = scan.nextLine();
        String str2 = scan.nextLine();
        scan.close();

        String[] snums1 = str1.split(" ");
        String[] snums2 = str2.split(" ");

        int[] nums1 = new int[snums1.length];
        int[] nums2 = new int[snums2.length];

        for(int i = 0; i < snums1.length; i++) {
            nums1[i] = Integer.parseInt(snums1[i]);
        }

        for(int i = 0; i < snums2.length; i++) {
            nums2[i] = Integer.parseInt(snums2[i]);
        }

        System.out.println(new Solution3().findMedianSortedArrays(nums1, nums2));

    }
}

File: Leetcode100/leetcode100_python/MedianofTwoSortedArrays3.py
from typing import List

class Solution:
    def getKthElement(self, nums1, nums2, k):
        if not nums1:
            return nums2[k - 1]
        if not nums2:
            return nums1[k - 1]
        if k == 1:
            return min(nums1[0], nums2[0])

        half = k // 2
        i1 = min(half, len(nums1)) - 1
        i2 = min(half, len(nums2)) - 1
        if nums1[i1] <= nums2[i2]:
            return self.getKthElement(nums1[i1 + 1 : ], nums2, k - i1 - 1)
        else:
            return self.getKthElement(nums1, nums2[i2 + 1 : ], k - i2 - 1)

        

    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        if (m + n) % 2 == 1:
            return self.getKthElement(nums1, nums2, (m + n) // 2 + 1)
        else:
            return (self.getKthElement(nums1, nums2, (m + n) // 2) + 
                self.getKthElement(nums1, nums2, (m + n) // 2 + 1)) / 2

if __name__ == "__main__":
    nums1 = [int(num) for num in input().split()]
    nums2 = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.findMedianSortedArrays(nums1, nums2))